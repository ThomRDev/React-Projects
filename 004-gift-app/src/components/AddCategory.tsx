import React, { ChangeEvent,FormEvent, useState } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

interface IPropsAddCategory{
    setValues : Function
}

const FormSearchCategory = styled.form`
    width: 100%;
    margin: 1em auto;
    display: flex;
`

const InputSearch = styled.input`
    width: 100%;
    margin: auto;
    font-size: 1.2em;
    padding: 1em;
    border: none;
    color: #ccc;
    appearance: none;
    background-color: #4A5E6D;
    border-radius: 0.25rem;
    transition: border-color .15s ease-in-out,box-shadow .15s ease-in-out;
    :focus{
        border-color: hsl(215.5deg, 95%, 90%);
        outline: 0;
        box-shadow: 0 0 0 0.2rem #0D161B;
    }
    ::placeholder{
        color: #A1A1A1;
    }
`
InputSearch.displayName="input"
FormSearchCategory.displayName="form"

const AddCategory = ({ setValues } : IPropsAddCategory) => {
    const [inputValue, setInputValue] = useState("")
    const handleSubmit = (event:FormEvent<HTMLFormElement>) =>{
        event.preventDefault()
        const data = inputValue.trim() 
        if(data.length){
            setValues((old:any)=>[data,...old])
            setInputValue("")
        }
    }
    const handleChange = (event:ChangeEvent<HTMLInputElement>) =>{
        event.preventDefault()
        const data = event.target.value
        setInputValue(data)
    }
    return (
        <FormSearchCategory onSubmit={handleSubmit} className="l-container">
            <InputSearch placeholder='Look for what you want, USA, Marines, etc' value={inputValue} onChange={handleChange} type="text" />
        </FormSearchCategory>
    )
}

AddCategory.propTypes = {
    setValues : PropTypes.func.isRequired
}

export default AddCategory
// https://stackoverflow.com/questions/55465673/jest-enzyme-with-styled-components