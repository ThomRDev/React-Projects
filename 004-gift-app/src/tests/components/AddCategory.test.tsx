import { shallow } from "enzyme"
import path from "path"

import AddCategory from "../../components/AddCategory"

describe(`Tests in ${__filename} file`, () => { 
    // es una simple funcion 
    // nos dira como fue llamada, si fue llamada o cuantas veces fue llamada
    const setValues = jest.fn()

    // tambien puedo crear esto
    // esto no nos dira nada
    // const setValues = ()=>{}

    const wrapper = shallow(<AddCategory setValues={setValues} />)
    
    test('Must render the component <AddCategory /> correctly', () => {

        // console.log(wrapper.html())
        expect(wrapper).toMatchSnapshot()
    })

    test('must change the input value', () => {
        const input = wrapper.find("input")
        // console.log(input.html());

        // simulando el evento onChange
        // pero como no esta en un entorno web hay propeidades que no entendera
        // el {} seria el evento
        input.simulate("change",{
            preventDefault : jest.fn(),
            // preventDefault(){
                // tambien puedo usar esto
            // },
            target:{
                value : "Data"
            }
        })

        // No funciona 
        // ya que lo captura en ese momento
        // console.log(input.html());
        // expect(input.prop("value")).toBe("Data")

        // si funciona
        // console.log(wrapper.find("input").html());
        expect(wrapper.find("input").prop("value")).toBe("Data")
    })

    test('the hook (setValues) must not be called', () => {
        // input vacio
        wrapper.find("input").simulate("change",{
            preventDefault : jest.fn(),
            target:{
                value : ""
            }
        })
        // tenemos que hacer lo anterior ya que las simulaciones si se guardan
        // tendremos que el wrapper y la funcion esten en un beforeEach

        wrapper.find("form").simulate("submit",{ preventDefault(){} })
        expect(setValues).not.toHaveBeenCalled()
    })
})


describe(`Tests in ${path.basename(__filename)} file`, () => {
    let setValues = jest.fn()
    let wrapper = shallow( <AddCategory setValues={setValues}  /> )
    beforeEach(()=>{
        // setValues = jest.fn()
        // tambien puedo limp√≠ar
        jest.clearAllMocks() // limpiar todo
        // o solo limiar la funcion
        // setValues.mockClear()
        wrapper = shallow( <AddCategory setValues={setValues}  /> )
    })
    test('Must render the component <AddCategory /> correctly', () => {
        expect(wrapper).toMatchSnapshot()
    })
    test('must change the input value', () => {
        const input = wrapper.find("input")
        input.simulate("change",{
            preventDefault : jest.fn(),
            target:{
                value : "Data"
            }
        })
        expect(wrapper.find("input").prop("value")).toBe("Data")
    })
    test('the hook (setValues) must not be called', () => {
        wrapper.find("form").simulate("submit",{ preventDefault(){} })
        expect(setValues).not.toHaveBeenCalled()
    })
    test('the hook (setValues) must be called', () => {
        wrapper.find("input").simulate("change",{
            preventDefault(){},
            target:{ value : "Data" }
        })
        wrapper.find("form").simulate("submit",{ preventDefault(){} })
        expect(setValues).toHaveBeenCalled()
    })
    test('submit event', () => {
        wrapper.find("input").simulate("change",{
            preventDefault(){},
            target:{ value : "asdasd" }
        })
        expect(setValues).toHaveBeenCalled()
        expect(setValues).toHaveBeenCalledWith(expect.any(Function))

        
        // expect(setValues).toHaveBeenCalledTimes(1) // que se llame 1 ves

        // Verificara que setValues ejecute una funcion
        // es decir que setValues reciba una funcion
        // expect(setValues).toHaveBeenCalledWith(function(){})
        // mejor
        // expect(setValues).toHaveBeenCalledWith(expect.any(Function))

        expect(wrapper.find("input").prop("value")).toBe("asdasd")
        wrapper.find("input").simulate("change",{
            preventDefault(){},
            target:{ value : "" }
        })
        expect(wrapper.find("input").prop("value")).toBe("")
        wrapper.find("form").simulate("submit",{ preventDefault(){} })
        expect(setValues).not.toHaveBeenCalled()
    })
})