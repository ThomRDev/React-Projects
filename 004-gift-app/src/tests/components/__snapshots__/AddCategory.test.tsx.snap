// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests in AddCategory.test.tsx file Must render the component <AddCategory /> correctly 1`] = `
<form
  className="l-container"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Look for what you want, USA, Marines, etc"
    type="text"
    value=""
  />
</form>
`;

exports[`Tests in C:\\Users\\ThomMR\\Desktop\\Learning-React-2022\\004-gift-app\\src\\tests\\components\\AddCategory.test.tsx file Must render the component <AddCategory /> correctly 1`] = `
<form
  className="l-container"
  onSubmit={[Function]}
>
  <input
    onChange={[Function]}
    placeholder="Look for what you want, USA, Marines, etc"
    type="text"
    value=""
  />
</form>
`;
